examples/classifier.c

```diff
@@ -15,6 +16,324 @@ float *get_regression_values(char **labels, int n)
+    save_weights(net, buff);
+    pthread_join(load_thread, 0);
+
+    free_network(net);
+    if(labels) free_ptrs((void**)labels, classes);
+    free_ptrs((void**)paths, plist->size);
+    free_list(plist);
+    //free(valid_accs);
+    //free(train_accs);
+    fclose(training_log);
+    //free(base);
+}
+
+
+
+
+float** calculate_confusion_matrix (char *datacfg, char *filename, char *weightfile, char* set_name)
+{
+    int i, j;
+    network *net = load_network(filename, weightfile, 0);
+    set_batch_network(net, 1);
+    srand(time(0));
+
+    list *options = read_data_cfg(datacfg);
+
+    char *label_list = option_find_str(options, "labels", "data/labels.list");
+    char *leaf_list = option_find_str(options, "leaves", 0);
+    if(leaf_list) change_leaves(net->hierarchy, leaf_list);
+    char *valid_list = option_find_str(options, set_name, "data/test.list");
+    int classes = option_find_int(options, "classes", 2);
+    int topk = option_find_int(options, "top", 1);
+
+    char **labels = get_labels(label_list);
+    list *plist = get_paths(valid_list);
+
+    char **paths = (char **)list_to_array(plist);
+    int m = plist->size;
+    free_list(plist);
+
+    float avg_acc = 0;
+    float avg_topk = 0;
+    int *indexes = calloc(topk, sizeof(int));
+
+    float** conf = malloc(classes * sizeof(float*));
+
+    for(i = 0; i < classes; ++i)
+        conf[i] = calloc(classes, sizeof(float));
+
+
+    for(i = 0; i < m; ++i){
+        int class = -1;
+        char *path = paths[i];
+        for(j = 0; j < classes; ++j){
+            if(strstr(path, labels[j])){
+                class = j;
+                break;
+            }
+        }
+        image im = load_image_color(paths[i], 0, 0);
+        image crop = center_crop_image(im, net->w, net->h);
+        
+        float *pred = network_predict(net, crop.data);
+
+        if(net->hierarchy) hierarchy_predictions(pred, net->outputs, net->hierarchy, 1, 1);
+
+        free_image(im);
+        free_image(crop);
+        top_k(pred, classes, topk, indexes);
+
+        if(indexes[0] == class) avg_acc += 1;
+
+        for(j = 0; j < topk; ++j)
+            if(indexes[j] == class) avg_topk += 1;
+
+        conf[class][indexes[0]]++;
+    }
+
+    free(indexes);
+    free_network(net);
+    free(paths);
+    
+    return conf;
+}
+
+
+
+float calculate_metric (float num, float den)
+{
+    return fabs(den) < 1e-8 ? 0.0 : num / den;
+}
+
+
+float calculate_accuracy (float** conf, int classes)
+{
+    int i, j;
+    float sum = 0.0, diag = 0.0;
+
+    for(i = 0; i < classes; ++i)
+    {
+        for(j = 0; j < classes; ++j)
+            sum += conf[j][i];
+
+        diag += conf[i][i];
+    }
+
+    return diag / sum;
+}
+
+
+float calculate_precision (float** conf, int classes)
+{
+    int i, j;
+    float mean = 0.0;
+
+    for(i = 0; i < classes; ++i)
+    {
+        float sum = 0.0;
+
+        for(j = 0; j < classes; ++j)
+            sum += conf[j][i];
+
+        mean += (conf[i][i] / sum);
```

Comment:
I think precision should be TP/(TP+FP), but this seems to be computing a normalized version of the accuracy... could you double check this?